[droidmedia] allow create droidmediabuffer from other formats

peat-psuwit opened the pull request:
droid_media_buffer_create_from_raw_data assumes that the data is always
YV12. When other format is passed in, it can cause a crash.
This commit make sure that the function uses the correct number of lines
and stride for the format passed in.

This can happen if Gstreamer happens to negotiate raw NV21 from
droidcamsrc to droideglsink because the elements in the middle doesn't
support ENCODED format passthrough. While the proper fix needs to be
in Gstreamer (see [1]), this patch is still needed for correctness.

This commit also add a logging message if calculated stride is too much
off. This can helps if the program crash because of this.

[1] https://gitlab.freedesktop.org/gstreamer/gst-plugins-base/merge_requests/467

peat-psuwit commented:
Ping the developer. @abranson?

abranson commented:
Good ping, thanks I hadn't seen this. Will take a look. It makes sense, but what's happening on your side that needs it?

monich commented on droidmediabuffer.cpp:
```diff
@@ -138,31 +140,73 @@ DroidMediaBuffer *droid_media_buffer_create_from_raw_data(uint32_t w, uint32_t h
+    lineWidthUV = w;
+    dstStrideUV = dstStrideY;
+    ALOGW("Encountered HAL_PIXEL_FORMAT_YCbCr_422_I. Please check if these values make sense."
+          "(format = %d, dstStrideUV = %d, strideUV = %d)", format, dstStrideUV, strideUV);
```
Missing break. And I generally don't like adding "I'm not sure how this is supposed to work" kind of code.

peat-psuwit replied:
Could you please point out where I can find out the correct values for this format?

peat-psuwit replied:
I've fixed that now. Please take another look.

peat-psuwit commented:
@abranson Sorry for the late reply.

> It makes sense, but what's happening on your side that needs it?

I'm trying to make gst-droid works on Sony Xperia X (though it's on Ubuntu Touch, not Mer/SailfishOS). On that device, it reports the viewfinder format as `0x7f000789` (COLOR_FormatSurface), which droidcamsrc promptly converts to Gst's ENCODED format.

When camerabin element tries to connect droidcamsrc's vfsrc pad to the sink, it'll insert a videoconvert element and a videoscale element (among others) between them. The issue is, videoconvert and videoscale doesn't support ENCODED format in its caps template. In this case, the caps negotiation process will then choose the failback NV21 format available, which translates to droidcamsrc's raw viewfinder mode.

Now, raw (system memory) buffers are produced from droidcamsrc, through videoconvert & videoscale, to droideglsink. In this case, droideglsink needs to create droidmediabuffer from these raw data (remember, they're in NV21 format). However, without this patch, it will read past the end of the buffer as NV21 has only half the lines in the UV plane compared to (assumed) YV12 format, resulting in SIGSEGV. With this patch, droideglsink will successfully create the droidmediabuffer and can display it. Even though the video recording isn't possible, at least there's no SIGSEGV.

----------------------------------------------------------------------------------------

As I said in the commit message, indeed the actual fix should be in the Gstreamer. However, if some other people use this to create droidmediabuffer from this or other formats (or if some encoded format gives NV21 out), it shouldn't SIGSEGV. That's why I submitted this patch.

abranson commented:
Interesting, but what's puzzling me is why this happens for you when everything works fine in Sailfish on the Xperia X. According to my dumped graph, vfsrc is sending NV12_64Z32 format data to the droideglsink inside the GstViewfinderBin, but there's no videoconvert or videoscale, just a queue and capsfilter.

It would surely be much less taxing for the device to have opaque buffers going through to the viewfinder though. Best of luck with that patch. But I think your problem with the viewfinder lies elsewhere.

peat-psuwit commented:
Hmm, I can't find any reference for that color in droidcamsrc's [format conversion table](https://github.com/sailfishos/gst-droid/blob/master/gst/droidcamsrc/gstdroidcamsrcdev.c#L575-L588) (I found it in droidvdec though). Do you have any special patch for that? However, if videoscale & videoconvert don't appear in your pipeline, then it should work just fine. (And BTW what is its systemMemory feature? If it's DroidMediaBuffer than the actual format doesn't matter.)

For the software video codecs, it depends on what color format that codec is outputting. What are the codec's caps?

abranson commented:
No there's no special patch. I'm doubting that's actually what's going through now - those .dot generated graphs can be tricky to read :)

peat-psuwit commented:
Ok, new version pushed. With some research, I've found that YCbCr_422_I is not a planar format. So, I fixed that. I also fixed the line widths for RGB formats and YCbCr_422_I, as they're not equal to the width.

And as a safety measure, I also added a check to make sure only the supported format is passed in before doing anything further.

abranson commented on droidmediabuffer.cpp:
```diff
@@ -114,10 +114,31 @@ DroidMediaBuffer *droid_media_buffer_create_from_raw_data(uint32_t w, uint32_t h
+  case HAL_PIXEL_FORMAT_YV12:
+  case HAL_PIXEL_FORMAT_YCbCr_422_SP:
+  case HAL_PIXEL_FORMAT_YCrCb_420_SP:
+  case /* QOMX_COLOR_FormatYUV420PackedSemiPlanar32m */ 0x7FA30C04:
```
These two are set in droidmediaconstants already. Would be better to use that rather than define them twice.

peat-psuwit replied:
Unfortunately, they're passed to gst-droid side without a way to retrieve it from the other part of droidmedia.

https://github.com/sailfishos/droidmedia/blob/52143f26ff6dc2346cf4f14792c175493cf529b1/droidmediaconstants.cpp#L80

The `c` variable here is passed from gst-droid and is stored nowhere within droidmedia.

abranson replied:
Well, there must be some way to define these in one place then use them everywhere. They occur three times each in this file.

peat-psuwit replied:
Hmm, what's the earliest version of Android droidmedia has to support? These 2 variables seem to be defined in AOSP's OMX_IVCommon.h since KitKat.

https://android.googlesource.com/platform/frameworks/native/+/kitkat-release/include/media/openmax/OMX_IVCommon.h#162

peat-psuwit replied:
And, in fact, will we ever have to create those vendor-specific color formats?

abranson replied:
Yeah, they were added because some Qualcomm devices use them for camera buffers iirc. Droidmedia still supports the Jolla1 with its 4.1 Android base, but MTK devices are also missing these defines.

abranson commented on droidmediabuffer.cpp:
```diff
@@ -138,31 +159,83 @@ DroidMediaBuffer *droid_media_buffer_create_from_raw_data(uint32_t w, uint32_t h
+  case HAL_PIXEL_FORMAT_RGBA_8888:
+  case HAL_PIXEL_FORMAT_RGBX_8888:
+  case HAL_PIXEL_FORMAT_BGRA_8888:
+    lineWidthY = 4 * w;
```
I've been testing this with RGBA, and it seems that the dstStrideY also needs multiplying by the bytes per pixel.

peat-psuwit replied:
Will fix that, but does the srideY (passed from gst-droid) have to be multiplied too?

abranson replied:
GST_VIDEO_INFO_COMP_STRIDE seems to already return the stride in bytes there.

abranson commented on droidmediabuffer.cpp:
```diff
@@ -138,31 +159,83 @@ DroidMediaBuffer *droid_media_buffer_create_from_raw_data(uint32_t w, uint32_t h
+    break;
+  default: // RGB, and YCbCr_422_I
+    if (strideUV != 0)
+      ALOGW("Get unexpected 2nd plane for the format that shouldn't have one.");
```
In this case, the global memcpy below should still happen if the caller erroneously passed a strideUV for an RGB format.

peat-psuwit replied:
I've made it set the `strideUV` to 0 in this case. If `strideY` equals to `dstStrideY`, the global `memcpy` should still happen.

peat-psuwit commented:
After another thought, I've decided to remove the support for 2 QCOM formats. Software codec shouldn't produce those formats, right?

abranson commented:
Hi, I've just merged a propose-allocation function into droideglsink, which means it can hand out droidmediabuffers to other gstreamer elements and we won't have to do this manual copying any more (and also that colour problem is solved). Could you try that out and see if it helps solve your problem?

peat-psuwit commented:
Sorry for the delay. I was unable to test it (or find a people to test it) on the original device for a while. Today I'm able to test it on the original device, and yes, it seems like the purpose-allocation approach works in my case. So, if you like, you can close this PR.

abranson commented:
Thanks. I think that function's unused now and will remove it at a later date.
